### PR TITLE
Fix join on arrays of unhashable types and allow hash join on all types supported at run-time

### PR DESCRIPTION
**Pull request by findepi**
## Which issue does this PR close?

None

## Rationale for this change

fix potential query failure due to incorrect `can_hash` result

## What changes are included in this PR?

- update `can_hash` 
- cleanup `compute_hashes`

## Are these changes tested?

Yes, unit tests.
query-based tests are hard to write -- it is unclear how to construct Union or RunEndEncoded types in SQL

## Are there any user-facing changes?

yes

### CODE REVIEW

File: datafusion/common/src/hash_utils.rs [hunk 1]
```diff
@@ -392,14 +392,6 @@ pub fn create_hashes<'a>(
                 let array: &FixedSizeBinaryArray = array.as_any().downcast_ref().unwrap();
                 hash_array(array, random_state, hashes_buffer, rehash)
             }
-            DataType::Decimal128(_, _) => {
```
Comment (Dandandan) [inline datafusion/common/src/hash_utils.rs:395]: Why can this be removed?

Comment (findepi) [reply datafusion/common/src/hash_utils.rs:395]: quoting commit message from the first commit https://github.com/apache/datafusion/pull/13388/commits/1e6994688b8c3eb814336ebe21e43dea79960b30

```
Remove unused code paths from create_hashes

The `downcast_primitive_array!` macro handles all primitive types
and only then delegates to fallbacks. It handles Decimal128 and
Decimal256 internally.
```

**Comment by alamb** [comment]
> query-based tests are hard to write -- it is unclear how to construct Union or RunEndEncoded types in SQL

The way to do so would be via `arrow_cast` 


However, those types don't seem to be supported via `arrow_cast` yet:
```
> select arrow_cast(1, 'Union(Union))');
Error during planning: Unsupported type 'Union(Union))'. Must be a supported arrow type name such as 'Int32' or 'Timestamp(Nanosecond, None)'. Error unrecognized word: Union
```

**Comment by alamb** [comment]
The other way to do it is to programatically setup the sqllogictest, as is done for `map.slt` (which didn't have any way to create `MapArray`s when it was written): https://github.com/apache/datafusion/blob/ac79ef3442e65f6197c7234da9fad964895b9101/datafusion/sqllogictest/src/test_context.rs#L84-L85

**Comment by findepi** [comment]
This change has two aspects

1. fixing hash joins on types that are not hashable, but can_hash considered them hashable. This is eg List of something unhashable. I couldn't write a test for this, since types such as listview or ree hit other problems.
2. unblocking hash joins on types that are actually hashable, but can_hash considered them not hashable. This is not as edge-case'y, for example `binary` is hashable, but we didn't use hash joins when joining on binary. I added a plan test for this.

**Comment by findepi** [comment]
@alamb please take another look

File: datafusion/sqllogictest/src/test_context.rs [hunk 2]
```diff
@@ -251,8 +253,11 @@ pub async fn register_table_with_many_types(ctx: &SessionContext) {
         .unwrap();
     ctx.register_catalog("my_catalog", Arc::new(catalog));
 
-    ctx.register_table("my_catalog.my_schema.t2", table_with_many_types())
-        .unwrap();
+    ctx.register_table(
```
Comment (alamb) [inline datafusion/sqllogictest/src/test_context.rs:256]: 👍

**Review by alamb — APPROVED** [review]
Thank you @findepi  -- I think this is a nice improvement and cleanup

**Pull request merged**